reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,428 @@
+# Revamping APIs for Offline Batch Inference
+
+## Summary
+Reduce the friction for batch inference of pre-trained models by updating the APIs to do the following:
+1. Remove AIR Checkpoints from the critical path for batch prediction.
+2. Simpler interface for chaining multiple preprocessing and prediction stages.
+
+### General Motivation
+There are changes to Ray Datasets in 2.3/2.4:
+1. Ray Datasets became lazy by default in Ray 2.3. Execution is no longer triggered until consumption is explicitly called.
+2. In Ray 2.4, Ray Datasets will move to streaming execution by default, resulting in there no longer being a separation between `Dataset` and `DatasetPipeline`.
+
+With these changes, certain API choices that were made for the AIR batch prediction API are no longer the most optimal for simplicity and usability.
+1. Previously, AIR would manually fuse preprocessing and prediction stages, requiring `BatchPredictor` to be aware of the preprocessors passed in via a `Checkpoint`. This is not needed anymore with Datasets lazy by default, as the fusion logic can be pushed down to the Datasets layer.
+2. Previously, the separation of `Dataset` and `DatasetPipeline` resulted in `BatchPredictor` needing 2 analgous APIs: `predict` and `predict_pipelined`. This is not needed anymore with streaming as default execution.
+
+### Should this change be within `ray` or outside?
+main `ray` project. Changes are made to Ray Data and Ray AIR level.
+
+## Stewardship
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experience Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@ericl, @clarkzinzow, @c21, @gjoliver
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@ericl
+
+## Design and Architecture
+
+## User facing API Changes
+
+1. Preprocessors can be directly passed to `Dataset.map_batches`.
+
+```python
+ds = ray.data.read_images(...)
+preprocessor1 = BatchMapper(lambda x: x+1)
+preprocessor2 = TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224))
+ds.map_batches(preprocessor1).map_batches(preprocessor2)
+```
+
+2. Predictors can be directly passed to `Dataset.map_batches`
+
+```python
+ds = ray.data.read_images(...)
+model = resnet50()
+predictor = TorchPredictor(model=model)
+
+# Alternatively, create from AIR Checkpoint.
+# predictor = TorchPredictor.from_checkpoint(air_checkpoint)
+
+predictions = ds.map_batches(predictor, compute=ray.data.ActorPoolStrategy(size=8))
+```
+
+Exception is raised if ActorPoolStrategy is not specified.
```

Comment:
What is the rationale behind this? As a user I would prefer if I can do
```
predictions = ds.map_batches(predictor)
```
and it does the right thing. Why do I need to care about ActorPoolStrategy?

Reply:
The default `map_batches` API is TaskPoolStrategy.

If we want to make `ActorPoolStrategy` the default, then we'd have to introduce a new API (like `map_predictor`) which we decided not to do in order to have a single unified API for data processing.

Reply:
If a basic batch inference user needs to know about `ActorPoolStrategy` I think we are failing our users, it is not something the user should know about (this is an abstraction leak).

Reply:
@pcmoritz your recommendation is go with a separate `map_predictor` API?

Reply:
I don't think it's very easy to avoid this abstraction leak *entirely*. Here are a few things we've learned from previous users of this API:
- No one uses the highest level BatchPredictor API, which does try to hide most of underlying details.
- Some users have expressed that they prefer lower-level APIs (such as Ray core), because they understand better what Ray is doing under the hood.

As for creating a special `map_predictor()` method, I am OK with that, but we have gotten strong feedback that it's confusing why there are two very similar map methods. However, upon reflection we may be able to address this concern by explicitly documenting this as the syntactic sugar / simpler form of `map_batches()` for batch inference, similar to how `map` is sugar for `map_batches`.

Reply:
It's a non-trivial tradeoff. For now my preference is to go with `map_batches` with `ActorPoolStrategy`. If we see enough user feedback that a separate `map_predictor` is acceptable, we can do that down the road. I feel it is an incremental path. @pcmoritz what do you think?

Another reason why I prefer this sequence is, for really simple use inference use cases, "it works" without the `ActorPoolStrategy` argument, just slower. @amogkam what is the rough performance difference with vs. without it?

Reply:
Why not use `map_batches` and if the argument is a class (like an actor class or Predictor class), it will use the `ActorPoolStrategy`? I don't see the tradeoff here. It can have a reasonable default like "auto determine the number of actors" and if people want to override, they can specify the `ActorPoolStrategy`.

Reply:
There is the tradeoff between explicitness/understandability and ease of getting started.

If we go with auto determining the number of actors (perhaps with a warning/tip on how to configure it), most users will probably start with this code path.

However if we raise an error and tell the user to specify ActorPoolStrategy(size=N), this starts users down the more explicit and understandable code path, though it adds an additional onboarding step.

The current API is preferring the more explicit route, because (1) users have tended to prefer explicit / easy to understand solutions, and (2) autoscaling introduces performance questionmarks (e.g., slow scale up) and/or complexities (e.g., predictive autoscaling) that can trip up new users trying to get something basic to work.

Reply:
This REP is just being consistent with the current behavior (https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L342-L352) and is not making a stance on whether there should be a default path for ActorPoolStrategy.

Whether there should be a default path for ActorPoolStrategy is an important discussion to have, but is orthogonal to this REP. Let's continue this discussion as a separate change on #api-changes. If we do decide to have a default ActorPoolStrategy path, then this REP will be modified accordingly.

Reply:
In the `before` code, the user didn't need to specify a number of actors, and now they have to, that looks like a clear regression with this proposal to me.

How was the number of actors determined beforehand?

Reply:
I'm specifically referring to this proposal that you suggested

> Why not use map_batches and if the argument is a class (like a Predictor class or another class that gets called), it will use the ActorPoolStrategy?

which is a more broader change that goes beyond this REP